### ISSUE TITLE
It seems the result we get is not the same as the repo shows

### ISSUE THREAD

**Issue by linbeyoung** [FranxYao/Long-Context-Data-Engineering#3]
![image](https://github.com/FranxYao/Long-Context-Data-Engineering/assets/34389681/2adb7784-bba1-49b5-bd26-a3bd14296d0f)
this is the result we get with the code in this repo. we follow the readme step by step, making sure the environment, model and requirement are the same with the repo, but we are puzzled that we can not have the same score, especially at about 4k-tokens where the score is very low.
我们使用仓库的源代码，模型和环境进行复现，得到的结果是上面这张图，想请问一下可能是哪里出现问题？

**Comment by FranxYao** [comment]
Hi thanks for the interest! I wonder:
* If you could paste the output of the wrong predictions and let's take a look at what the model says?
* I was told that the model behavior may be sensitive to prompt. So maybe change the prompt at https://github.com/FranxYao/Long-Context-Data-Engineering/blob/main/eval/needle/needle_in_haystack.py#L201 to: 
```python
test_format=f"<|im_start|> This is a very long story book: <book> {context} </book>.\n Based on the content of the book, Question: {self.retrieval_question}\nAnswer: The best thing to do in San Francisco is"
```

**Comment by FranxYao** [comment]
OK try this branch which may fix the repeating issue at 4K length:
https://github.com/FranxYao/Long-Context-Data-Engineering/tree/fix_rope

Where the difference is at the following line:
https://github.com/FranxYao/Long-Context-Data-Engineering/blob/5b1fbd419e7a2ebc8644c7b78b8f7ed7186dcbe2/eval/needle/needle_in_haystack.py#L171

**Comment by linbeyoung** [comment]
[llama-2-7b-80k-result.zip](https://github.com/FranxYao/Long-Context-Data-Engineering/files/14353636/llama-2-7b-80k-result.zip)
this is the output.

**Comment by FranxYao** [comment]
<img width="960" alt="image" src="https://github.com/FranxYao/Long-Context-Data-Engineering/assets/17723677/c4972fb5-aa94-422a-96a9-88fd81b1418b">
Here is what I get from the new branch.

The model behavior is quite interesting though. If you could confirm you can get similar results I'll merge it to the main branch.

**Comment by linbeyoung** [comment]
> <img alt="image" width="960" src="https://private-user-images.githubusercontent.com/17723677/306484703-c4972fb5-aa94-422a-96a9-88fd81b1418b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0ODY0OTIsIm5iZiI6MTcwODQ4NjE5MiwicGF0aCI6Ii8xNzcyMzY3Ny8zMDY0ODQ3MDMtYzQ5NzJmYjUtYWE5NC00MjJhLTk2YTktODhmZDgxYjE0MThiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDAzMjk1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjYTViZDhlODMwODIzNmJkODEwZjdiMmQ2OWE5ZWJlODkwYjFiYmE1YzExMzFjNTFlZWRkOTNlYTI4ZjgyNDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BdrW89DPrxYcJTYVKgoi8s6x4z-Dmd__Yqrg1wu54hE"> Here is what I get from the new branch.
> The model behavior is quite interesting though. If you could confirm you can get similar results I'll merge it to the main branch.

Is it use the origin prompt?

**Comment by FranxYao** [comment]
Yes it does (thought would also be interesting to compare the two prompts)

**Comment by linbeyoung** [comment]
![企业微信截图_3a89a116-7dd2-41dc-a94f-4a879f65e210](https://github.com/FranxYao/Long-Context-Data-Engineering/assets/34389681/34a009bc-ce9e-43a4-a7a7-1efc1d9295d1)
We've implemented the new branch and observed improved results. Indeed, it's significantly better. However, the overall score is 0.848, which still presents a slight discrepancy compared to the results mentioned in your repository. Could there be any potential reasons we might have overlooked?

**Comment by rooa** [comment]
Hey @FranxYao , thanks for the great work on this paper. I was wondering, did you use the prompt in the code or the modified prompt above for the figure in the paper?

```python
# prompt in the code
f"<|im_start|> This is a very long story book: <book> {context} </book>.\n Based on the content of the book, Question: {self.retrieval_question}\nAnswer:"

# prompt suggested above
f"<|im_start|> This is a very long story book: <book> {context} </book>.\n Based on the content of the book, Question: {self.retrieval_question}\nAnswer: The best thing to do in San Francisco is"
```

Thanks!

**Comment by HannahBenita** [comment]
@[marcobellagente93](https://github.com/marcobellagente93) and I have noticed a similar inconsistency with the needle-in-a-haystack evaluation for another project.
We have tracked down the problem to the reading of PaulGraham essays with glob.glob which is non deterministic, therefore the files are loaded in arbitrary order leading to a different ‘{context}’ being inserted in the prompt. Which in turn then leads to different behaviour of the model. 
Interestingly this phenomenon only appears in different clones of the repo, for one repo the ‘{context}’ seems to be consistent. You can easily verify this behaviour by cloning the Needle-in-a-haystack repo twice and printing the first 200 characters of the context in the read_context_files function. 

For https://arxiv.org/abs/2402.17834 we therefore opted to report the mean and std across 10 runs with random context.

**Comment by rooa** [comment]
Great find! I totally missed that. File order being different after every cloning but consistent within each clone is probably because files are copied and registered by the file system arbitrarily.

**Comment by FranxYao** [comment]
> @[marcobellagente93](https://github.com/marcobellagente93) and I have noticed a similar inconsistency with the needle-in-a-haystack evaluation for another project. We have tracked down the problem to the reading of PaulGraham essays with glob.glob which is non deterministic, therefore the files are loaded in arbitrary order leading to a different ‘{context}’ being inserted in the prompt. Which in turn then leads to different behaviour of the model. Interestingly this phenomenon only appears in different clones of the repo, for one repo the ‘{context}’ seems to be consistent. You can easily verify this behaviour by cloning the Needle-in-a-haystack repo twice and printing the first 200 characters of the context in the read_context_files function.
> 
> For https://arxiv.org/abs/2402.17834 we therefore opted to report the mean and std across 10 runs with random context.

Thanks for the meticulous study!! I guess this may contribute to the reason.

Also I have updated the code and merged the `fix_rope` branch. The model should give more stable output now
